### PR TITLE
use alignment provided by elf for note segments

### PR DESCRIPTION
**Pull request by name1e5s**
On some android version(more precisely, 6.0 to 9.0, according to [this commit](https://github.com/aosp-mirror/platform_bionic/commit/e09ab2cc2ff41ce7519c143d448371ddd62afc59) and [this commit](https://github.com/aosp-mirror/platform_bionic/commit/12533271fbc1eec185943691fa8a4410d512de89#diff-533086407911d0a23a26d0d51378f0067a507439bb838fef299429a425add1b6)), google sets alignment of its brand section `.note.android.ident`  to 2. Which may cause `findshlibs` panic when we try to get id for a `so`. I think it's better to use alignment provided by elf itself.

### CODE REVIEW

File: src/linux/mod.rs [hunk 1]
```diff
@@ -88,17 +89,12 @@ impl<'a> Segment<'a> {
...
+        let is_note = self.is_note();
         // `man 5 readelf` says that all of the `Nhdr`, name, and descriptor are
-        // always 4-byte aligned, but we copy this alignment behavior from
-        // `readelf` since that seems to match reality in practice.
-        let alignment = std::cmp::max(self.phdr().p_align as usize, 4);
+        // always 4-byte aligned, but in practice android set its `.note.android.ident`
+        // alignment to 2 on some versions, we'd better use alignment provided by elf.
+        let alignment = std::cmp::max(self.phdr().p_align as usize, 1);
         let align_up = move |data: &'a [u8]| {
-            if alignment != 4 && alignment != 8 {
```
Comment (name1e5s) [inline src/linux/mod.rs:98]: ELF spec forces alignment to be power of 2, it ok to not check alignment here.

**Comment by philipc** [comment]
> Which may cause findshlibs panic when we try to get id for a so.

Where is the panic?

> I think it's better to use alignment provided by elf itself.

The problem is that some linux binaries set the alignment to 0 or 1, but still actually use a 4 byte alignment within the notes.

**Comment by name1e5s** [comment]
> > Which may cause findshlibs panic when we try to get id for a so.
> 
> Where is the panic?

When parse a note segment with only .note.android.ident section, and that section is placed with an offset can’t divide by 4. In fact I encountered this panic last month when I test findshlibs with an older version of android NDK. I’ll try to reproduce the panic.

> > I think it's better to use alignment provided by elf itself.
> 
> The problem is that some linux binaries set the alignment to 0 or 1, but still actually use a 4 byte alignment within the notes.

Maybe we can have a special process for .note.android.ident on android platform only, and keep things unchanged on other linux platforms?

**Comment by name1e5s** [comment]
> > Which may cause findshlibs panic when we try to get id for a so.
> 
> Where is the panic?
> 
> > I think it's better to use alignment provided by elf itself.
> 
> The problem is that some linux binaries set the alignment to 0 or 1, but still actually use a 4 byte alignment within the notes.

I find that method `notes` is only used for get `id` for a `so`, may be we can simply return an empty iterator instead of panic the whole thread using `assert_eq!(data.as_ptr() as usize % alignment, 0)` when pointer alignment doesn't meet our requirements.
Will open a new PR to do this.

**Pull request closed without merging**